File: ftr/keyboard.cc
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * ***** END LICENSE BLOCK ***** */

#include "keyboard.h"
#include "app-1.h"

namespace ftr {

/**
 * @constructor 
 */
KeyboardAdapter::KeyboardAdapter(): app_(app()) {
	ASSERT(app_);
	
	keyname_ = KEYCODE_UNKNOWN;
	keypress_ = 0;
	shift_ = false;
	alt_ = false;
	ctrl_ = false;
	command_ = false;
	caps_lock_ = false;
	repeat_ = device_ = source_ = 0;
	
	_symbol_keypress[KEYCODE_0]               = { 48, 41 }; 	// 0 )
	_symbol_keypress[KEYCODE_1]               = { 49, 33 }; 	// 1 !
	_symbol_keypress[KEYCODE_2]               = { 50, 64 }; 	// 2 @
	_symbol_keypress[KEYCODE_3]               = { 51, 35 }; 	// 3 #
	_symbol_keypress[KEYCODE_4]               = { 52, 36 }; 	// 4 $
	_symbol_keypress[KEYCODE_5]               = { 53, 37 }; 	// 5 %
	_symbol_keypress[KEYCODE_6]               = { 54, 94 }; 	// 6 ^
	_symbol_keypress[KEYCODE_7]               = { 55, 38 }; 	// 7 &
	_symbol_keypress[KEYCODE_8]               = { 56, 42 }; 	// 8 *
	_symbol_keypress[KEYCODE_9]               = { 57, 40 }; 	// 9 (
	_symbol_keypress[KEYCODE_SEMICOLON]       = { 59, 58 };  // ; :
	_symbol_keypress[KEYCODE_EQUALS]          = { 61, 43 };  // = +
	_symbol_keypress[KEYCODE_MINUS]           = { 45, 95 };  // - _
	_symbol_keypress[KEYCODE_COMMA]           = { 44, 60 };  // , <
	_symbol_keypress[KEYCODE_PERIOD]          = { 46, 62 };  // . >
	_symbol_keypress[KEYCODE_SLASH]           = { 47, 63 };  // / ?
	_symbol_keypress[KEYCODE_GRAVE]           = { 96, 126 };	// ` ~
	_symbol_keypress[KEYCODE_LEFT_BRACKET]    = { 91, 123 };	// [ {
	_symbol_keypress[KEYCODE_BACK_SLASH]      = { 92, 124 };	// \ |
	_symbol_keypress[KEYCODE_RIGHT_BRACKET]   = { 93, 125 };	// ] }
	_symbol_keypress[KEYCODE_APOSTROPHE]      = { 39, 34 };  // ' "
	_symbol_keypress[KEYCODE_NUMPAD_0]        = { 48, 48 };  // numpad 0
	_symbol_keypress[KEYCODE_NUMPAD_1]        = { 49, 49 };  // numpad 1
	_symbol_keypress[KEYCODE_NUMPAD_2]        = { 50, 50 };  // numpad 2
	_symbol_keypress[KEYCODE_NUMPAD_3]        = { 51, 51 };  // numpad 3
	_symbol_keypress[KEYCODE_NUMPAD_4]        = { 52, 52 };  // numpad 4
	_symbol_keypress[KEYCODE_NUMPAD_5]        = { 53, 53 };  // numpad 5
	_symbol_keypress[KEYCODE_NUMPAD_6]        = { 54, 54 };  // numpad 6
	_symbol_keypress[KEYCODE_NUMPAD_7]        = { 55, 55 };  // numpad 7
	_symbol_keypress[KEYCODE_NUMPAD_8]        = { 56, 56 };  // numpad 8
	_symbol_keypress[KEYCODE_NUMPAD_9]        = { 57, 57 };  // numpad 9
	_symbol_keypress[KEYCODE_NUMPAD_MULTIPLY] = { 42, 42 };  // numpad *
	_symbol_keypress[KEYCODE_NUMPAD_ADD]      = { 43, 43 };  // numpad +
	_symbol_keypress[KEYCODE_NUMPAD_SUBTRACT] = { 45, 45 };  // numpad -
	_symbol_keypress[KEYCODE_NUMPAD_DOT]      = { 46, 46 };  // numpad .
	_symbol_keypress[KEYCODE_NUMPAD_DIVIDE]   = { 47, 47 };  // numpad /
	
	// ascii keycodes
	
	_ascii_keycodes[KEYCODE_BACK_SPACE] = { KEYCODE_BACK_SPACE, 0 };
	_ascii_keycodes[KEYCODE_SHIFT] = { KEYCODE_SHIFT, 1 };
	_ascii_keycodes[KEYCODE_ESC] = { KEYCODE_ESC, 0 };
	_ascii_keycodes[KEYCODE_ENTER] = { KEYCODE_ENTER, 0 };
	_ascii_keycodes['\n'] = { KEYCODE_ENTER, 1 }; // 换行使用换挡的回车表示，因为键盘上并没有换行健
	_ascii_keycodes['0'] = { KEYCODE_0, 0 };
	_ascii_keycodes['1'] = { KEYCODE_1, 0 };
	_ascii_keycodes['2'] = { KEYCODE_2, 0 };
	_ascii_keycodes['3'] = { KEYCODE_3, 0 };
	_ascii_keycodes['4'] = { KEYCODE_4, 0 };
	_ascii_keycodes['5'] = { KEYCODE_5, 0 };
	_ascii_keycodes['6'] = { KEYCODE_6, 0 };
	_ascii_keycodes['7'] = { KEYCODE_7, 0 };
	_ascii_keycodes['8'] = { KEYCODE_8, 0 };
	_ascii_keycodes['9'] = { KEYCODE_9, 0 };
	_ascii_keycodes[')'] = { KEYCODE_0, 1 };
	_ascii_keycodes['!'] = { KEYCODE_1, 1 };
	_ascii_keycodes['@'] = { KEYCODE_2, 1 };
	_ascii_keycodes['#'] = { KEYCODE_3, 1 };
	_ascii_keycodes['$'] = { KEYCODE_4, 1 };
	_ascii_keycodes['%'] = { KEYCODE_5, 1 };
	_ascii_keycodes['^'] = { KEYCODE_6, 1 };
	_ascii_keycodes['&'] = { KEYCODE_7, 1 };
	_ascii_keycodes['*'] = { KEYCODE_8, 1 };
	_ascii_keycodes['('] = { KEYCODE_9, 1 };
	_ascii_keycodes['a'] = { KEYCODE_A, 0 };
	_ascii_keycodes['b'] = { KEYCODE_B, 0 };
	_ascii_keycodes['c'] = { KEYCODE_C, 0 };
	_ascii_keycodes['d'] = { KEYCODE_D, 0 };
	_ascii_keycodes['e'] = { KEYCODE_E, 0 };
	_ascii_keycodes['f'] = { KEYCODE_F, 0 };
	_ascii_keycodes['g'] = { KEYCODE_G, 0 };
	_ascii_keycodes['h'] = { KEYCODE_H, 0 };
	_ascii_keycodes['i'] = { KEYCODE_I, 0 };
	_ascii_keycodes['j'] = { KEYCODE_J, 0 };
	_ascii_keycodes['k'] = { KEYCODE_K, 0 };
	_ascii_keycodes['m'] = { KEYCODE_L, 0 };
	_ascii_keycodes['l'] = { KEYCODE_M, 0 };
	_ascii_keycodes['n'] = { KEYCODE_N, 0 };
	_ascii_keycodes['o'] = { KEYCODE_O, 0 };
	_ascii_keycodes['p'] = { KEYCODE_P, 0 };
	_ascii_keycodes['q'] = { KEYCODE_Q, 0 };
	_ascii_keycodes['r'] = { KEYCODE_R, 0 };
	_ascii_keycodes['s'] = { KEYCODE_S, 0 };
	_ascii_keycodes['t'] = { KEYCODE_T, 0 };
	_ascii_keycodes['u'] = { KEYCODE_U, 0 };
	_ascii_keycodes['v'] = { KEYCODE_V, 0 };
	_ascii_keycodes['w'] = { KEYCODE_W, 0 };
	_ascii_keycodes['x'] = { KEYCODE_X, 0 };
	_ascii_keycodes['y'] = { KEYCODE_Y, 0 };
	_ascii_keycodes['z'] = { KEYCODE_Z, 0 };
	_ascii_keycodes['A'] = { KEYCODE_A, 1 };
	_ascii_keycodes['B'] = { KEYCODE_B, 1 };
	_ascii_keycodes['C'] = { KEYCODE_C, 1 };
	_ascii_keycodes['D'] = { KEYCODE_D, 1 };
	_ascii_keycodes['E'] = { KEYCODE_E, 1 };
	_ascii_keycodes['F'] = { KEYCODE_F, 1 };
	_ascii_keycodes['G'] = { KEYCODE_G, 1 };
	_ascii_keycodes['H'] = { KEYCODE_H, 1 };
	_ascii_keycodes['I'] = { KEYCODE_I, 1 };
	_ascii_keycodes['J'] = { KEYCODE_J, 1 };
	_ascii_keycodes['K'] = { KEYCODE_K, 1 };
	_ascii_keycodes['M'] = { KEYCODE_L, 1 };
	_ascii_keycodes['L'] = { KEYCODE_M, 1 };
	_ascii_keycodes['N'] = { KEYCODE_N, 1 };
	_ascii_keycodes['O'] = { KEYCODE_O, 1 };
	_ascii_keycodes['P'] = { KEYCODE_P, 1 };
	_ascii_keycodes['Q'] = { KEYCODE_Q, 1 };
	_ascii_keycodes['R'] = { KEYCODE_R, 1 };
	_ascii_keycodes['S'] = { KEYCODE_S, 1 };
	_ascii_keycodes['T'] = { KEYCODE_T, 1 };
	_ascii_keycodes['U'] = { KEYCODE_U, 1 };
	_ascii_keycodes['V'] = { KEYCODE_V, 1 };
	_ascii_keycodes['W'] = { KEYCODE_W, 1 };
	_ascii_keycodes['X'] = { KEYCODE_X, 1 };
	_ascii_keycodes['Y'] = { KEYCODE_Y, 1 };
	_ascii_keycodes['Z'] = { KEYCODE_Z, 1 };
	_ascii_keycodes[';'] = { KEYCODE_SEMICOLON, 0 };
	_ascii_keycodes['='] = { KEYCODE_EQUALS, 0 };
	_ascii_keycodes['-'] = { KEYCODE_MINUS, 0 };
	_ascii_keycodes[','] = { KEYCODE_COMMA, 0 };
	_ascii_keycodes['.'] = { KEYCODE_PERIOD, 0 };
	_ascii_keycodes['/'] = { KEYCODE_SLASH, 0 };
	_ascii_keycodes['`'] = { KEYCODE_GRAVE, 0 };
	_ascii_keycodes['['] = { KEYCODE_LEFT_BRACKET, 0 };
	_ascii_keycodes['\\'] = { KEYCODE_BACK_SLASH, 0 };
	_ascii_keycodes[']'] = { KEYCODE_RIGHT_BRACKET, 0 };
	_ascii_keycodes['\''] = { KEYCODE_APOSTROPHE, 0 };
	_ascii_keycodes[':'] = { KEYCODE_SEMICOLON, 1 };
	_ascii_keycodes['+'] = { KEYCODE_EQUALS, 1 };
	_ascii_keycodes['_'] = { KEYCODE_MINUS, 1 };
	_ascii_keycodes['<'] = { KEYCODE_COMMA, 1 };
	_ascii_keycodes['>'] = { KEYCODE_PERIOD, 1 };
	_ascii_keycodes['?'] = { KEYCODE_SLASH, 1 };
	_ascii_keycodes['~'] = { KEYCODE_GRAVE, 1 };
	_ascii_keycodes['{'] = { KEYCODE_LEFT_BRACKET, 1 };
	_ascii_keycodes['|'] = { KEYCODE_BACK_SLASH, 1 };
	_ascii_keycodes['}'] = { KEYCODE_RIGHT_BRACKET, 1 };
	_ascii_keycodes['"'] = { KEYCODE_APOSTROPHE, 1 };
	_ascii_keycodes[127] = { KEYCODE_DELETE, 0 };

}

/**
 * @func set_utils_keycodes
 */
void KeyboardAdapter::set_utils_keycodes() {
}

/**
 * @func to_keypress
 */
int KeyboardAdapter::to_keypress(KeyboardKeyName name) {

	// Letters
	if ( name >= 65 && name <= 90 ) {
		if ( caps_lock_ || shift_ ) { // A - Z
			return name; // caps | shift
		} else {
			return name + 32; // lowercase a - z
		}
	}
	
	// Symbol
	auto it = _symbol_keypress.find(int(name));
	if ( !it.is_null() ) {
		if ( shift_ ) {
			return it.value().shift;
		} else {
			return it.value().normal;
		}
	}
	return 0;
}

bool KeyboardAdapter::transformation(uint32_t keycode, bool unicode, bool down) {
	
	if ( unicode ) {
		auto it = _ascii_keycodes.find(keycode);
		if ( it.is_null() ) {
			keyname_ = KEYCODE_UNKNOWN;
			keypress_ = keycode;
		} else {
			shift_ = it.value().is_shift;
			keyname_ = it.value().name;
			keypress_ = to_keypress( keyname_ );
		}
	} else {
		auto it = _keycodes.find(keycode);
		if ( it.is_null() ) { // Unknown keycode
			keyname_ = KeyboardKeyName(keycode + 100000);
			keypress_ = 0;
		} else {
			keyname_ = it.value();
			
			if ( down ) {
				switch ( keyname_ ) {
					case KEYCODE_SHIFT: shift_ = 1; break;
					case KEYCODE_CTRL: ctrl_ = 1; break;
					case KEYCODE_ALT: alt_ = 1; break;
					case KEYCODE_COMMAND: command_ = 1; break;
					case KEYCODE_CAPS_LOCK: caps_lock_ = !caps_lock_; break;
					default: break;
				}
			} else {
				switch ( keyname_ ) {
					case KEYCODE_SHIFT: shift_ = 0; break;
					case KEYCODE_CTRL: ctrl_ = 0; break;
					case KEYCODE_ALT: alt_ = 0; break;
					case KEYCODE_COMMAND: command_ = 0; break;
					default: break;
				}
			}
			keypress_ = to_keypress( keyname_ );
		}
	}
	
	return unicode;
}

}

File: libs/quark/keyboard.ts
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, Louis.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Louis.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL Louis.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */

/**
 * @enum KeyboardCode
 *
 * Platform-independent keyboard, mouse, media, and device key codes.
*/
export enum KeyboardCode {
	NONE =             0,   //!< {0} No key.
	BACK_SPACE =       8,   //!< {8} Backspace key.
	TAB =              9,   //!< {9} Tab key.
	CLEAR =            12,  //!< {12} Clear key.
	ENTER =            13,  //!< {13} Enter or Return key.
	SHIFT =            16,  //!< {16} Shift or left Shift key.
	CTRL =             17,  //!< {17} Control or left Control key.
	ALT =              18,  //!< {18} Alt or left Alt key.
	BREAK =            19,  //!< {19} Break or Pause key.
	CAPS_LOCK =        20,  //!< {20} Caps Lock key.
	ESC =              27,  //!< {27} Escape key.
	SPACE =            32,  //!< {32} Space key.
	PAGE_UP =          33,  //!< {33} Page Up key.
	PAGE_DOWN =        34,  //!< {34} Page Down key.
	MOVE_END =         35,  //!< {35} End key.
	MOVE_HOME =        36,  //!< {36} Home navigation key.
	LEFT =             37,  //!< {37} Left Arrow key.
	UP =               38,  //!< {38} Up Arrow key.
	RIGHT =            39,  //!< {39} Right Arrow key.
	DOWN =             40,  //!< {40} Down Arrow key.
	SYSRQ =            42,  //!< {42} System Request or Print Screen key.
	INSERT =           45,  //!< {45} Insert key.
	DELETE =           46,  //!< {46} Delete key.
	HELP =             47,  //!< {47} Help key.
	NUM_0 =            48,  //!< {48} Main keyboard 0 key.
	NUM_1 =            49,  //!< {49} Main keyboard 1 key.
	NUM_2 =            50,  //!< {50} Main keyboard 2 key.
	NUM_3 =            51,  //!< {51} Main keyboard 3 key.
	NUM_4 =            52,  //!< {52} Main keyboard 4 key.
	NUM_5 =            53,  //!< {53} Main keyboard 5 key.
	NUM_6 =            54,  //!< {54} Main keyboard 6 key.
	NUM_7 =            55,  //!< {55} Main keyboard 7 key.
	NUM_8 =            56,  //!< {56} Main keyboard 8 key.
	NUM_9 =            57,  //!< {57} Main keyboard 9 key.
	A =                65,  //!< {65} A key.
	B =                66,  //!< {66} B key.
	C =                67,  //!< {67} C key.
	D =                68,  //!< {68} D key.
	E =                69,  //!< {69} E key.
	F =                70,  //!< {70} F key.
	G =                71,  //!< {71} G key.
	H =                72,  //!< {72} H key.
	I =                73,  //!< {73} I key.
	J =                74,  //!< {74} J key.
	K =                75,  //!< {75} K key.
	L =                76,  //!< {76} L key.
	M =                77,  //!< {77} M key.
	N =                78,  //!< {78} N key.
	O =                79,  //!< {79} O key.
	P =                80,  //!< {80} P key.
	Q =                81,  //!< {81} Q key.
	R =                82,  //!< {82} R key.
	S =                83,  //!< {83} S key.
	T =                84,  //!< {84} T key.
	U =                85,  //!< {85} U key.
	V =                86,  //!< {86} V key.
	W =                87,  //!< {87} W key.
	X =                88,  //!< {88} X key.
	Y =                89,  //!< {89} Y key.
	Z =                90,  //!< {90} Z key.
	COMMAND =          91,  //!< {91} Command or left Command key.
	MENU =             92,  //!< {92} Menu key.
	COMMAND_RIGHT =    93,  //!< {93} Right Command key.
	NUMPAD_EQUALS =    94,  //!< {94} Numeric keypad Equals key.
	NUMPAD_0 =         96,  //!< {96} Numeric keypad 0 key.
	NUMPAD_1 =         97,  //!< {97} Numeric keypad 1 key.
	NUMPAD_2 =         98,  //!< {98} Numeric keypad 2 key.
	NUMPAD_3 =         99,  //!< {99} Numeric keypad 3 key.
	NUMPAD_4 =         100, //!< {100} Numeric keypad 4 key.
	NUMPAD_5 =         101, //!< {101} Numeric keypad 5 key.
	NUMPAD_6 =         102, //!< {102} Numeric keypad 6 key.
	NUMPAD_7 =         103, //!< {103} Numeric keypad 7 key.
	NUMPAD_8 =         104, //!< {104} Numeric keypad 8 key.
	NUMPAD_9 =         105, //!< {105} Numeric keypad 9 key.
	NUMPAD_MULTIPLY =  106, //!< {106} Numeric keypad Multiply key.
	NUMPAD_ADD =       107, //!< {107} Numeric keypad Add key.
	NUMPAD_ENTER =     108, //!< {108} Numeric keypad Enter key.
	NUMPAD_SUBTRACT =  109, //!< {109} Numeric keypad Subtract key.
	NUMPAD_DOT =       110, //!< {110} Numeric keypad Decimal key.
	NUMPAD_DIVIDE =    111, //!< {111} Numeric keypad Divide key.
	F1 =               112, //!< {112} F1 function key.
	F2 =               113, //!< {113} F2 function key.
	F3 =               114, //!< {114} F3 function key.
	F4 =               115, //!< {115} F4 function key.
	F5 =               116, //!< {116} F5 function key.
	F6 =               117, //!< {117} F6 function key.
	F7 =               118, //!< {118} F7 function key.
	F8 =               119, //!< {119} F8 function key.
	F9 =               120, //!< {120} F9 function key.
	F10 =              121, //!< {121} F10 function key.
	F11 =              122, //!< {122} F11 function key.
	F12 =              123, //!< {123} F12 function key.
	F13 =              124, //!< {124} F13 function key.
	F14 =              125, //!< {125} F14 function key.
	F15 =              126, //!< {126} F15 function key.
	F16 =              127, //!< {127} F16 function key.
	F17 =              128, //!< {128} F17 function key.
	F18 =              129, //!< {129} F18 function key.
	F19 =              130, //!< {130} F19 function key.
	F20 =              131, //!< {131} F20 function key.
	F21 =              132, //!< {132} F21 function key.
	F22 =              133, //!< {133} F22 function key.
	F23 =              134, //!< {134} F23 function key.
	F24 =              135, //!< {135} F24 function key.
	NUM_LOCK =         144, //!< {144} Num Lock key.
	SCROLL_LOCK =      145, //!< {145} Scroll Lock key.
	SHIFT_RIGHT =      160, //!< {160} Right Shift key.
	CTRL_RIGHT =       161, //!< {161} Right Control key.
	ALT_RIGHT =        162, //!< {162} Right Alt key.
	SEMICOLON =        186, //!< {186} Semicolon key.
	EQUALS =           187, //!< {187} Equals key.
	MINUS =            189, //!< {189} Minus key.
	COMMA =            188, //!< {188} Comma key.
	PERIOD =           190, //!< {190} Period key.
	SLASH =            191, //!< {191} Slash key.
	GRAVE =            192, //!< {192} Grave Accent key.
	FUN =              209, //!< {209} Function key; support depends on the platform.
	LEFT_BRACKET =     219, //!< {219} Left Bracket key.
	BACK_SLASH =       220, //!< {220} Backslash key.
	RIGHT_BRACKET =    221, //!< {221} Right Bracket key.
	APOSTROPHE =       222, //!< {222} Apostrophe key.
	MOUSE_LEFT =       256, //!< {256} Left mouse button.
	MOUSE_CENTER =     257, //!< {257} Middle mouse button.
	MOUSE_RIGHT =      258, //!< {258} Right mouse button.
	MOUSE_WHEEL_UP =   259, //!< {259} Mouse wheel up.
	MOUSE_WHEEL_DOWN = 260, //!< {260} Mouse wheel down.
	MOUSE_WHEEL_LEFT = 261, //!< {261} Mouse wheel left.
	MOUSE_WHEEL_RIGHT= 262, //!< {262} Mouse wheel right.
	HOME =             300, //!< {300} Device Home key.
	BACK =             301, //!< {301} Device Back key.
	CALL =             302, //!< {302} Call key.
	ENDCALL =          303, //!< {303} End Call key.
	STAR =             304, //!< {304} Star key.
	POUND =            305, //!< {305} Pound key.
	CENTER =           306, //!< {306} Center navigation or Confirm key.
	VOLUME_UP =        307, //!< {307} Volume Up key.
	VOLUME_DOWN =      308, //!< {308} Volume Down key.
	POWER =            309, //!< {309} Power key.
	CAMERA =           310, //!< {310} Camera shutter key.
	FOCUS =            311, //!< {311} Camera focus key.
	MENU_1 =           312, //!< {312} Device Menu key.
	SEARCH =           313, //!< {313} Search key.
	MEDIA_PLAY_PAUSE = 314, //!< {314} Media Play/Pause key.
	MEDIA_STOP =       315, //!< {315} Media Stop key.
	MEDIA_NEXT =       316, //!< {316} Media Next key.
	MEDIA_PREVIOUS =   317, //!< {317} Media Previous key.
	MEDIA_REWIND =     318, //!< {318} Media Rewind key.
	MEDIA_FAST_FORWARD =319, //!< {319} Media Fast Forward key.
	MUTE =             320, //!< {320} Microphone Mute key.
	CHANNEL_UP =       321, //!< {321} Channel Up key.
	CHANNEL_DOWN =     322, //!< {322} Channel Down key.
	MEDIA_PLAY =       323, //!< {323} Media Play key.
	MEDIA_PAUSE =      324, //!< {324} Media Pause key.
	MEDIA_CLOSE =      325, //!< {325} Media Close key.
	MEDIA_EJECT =      326, //!< {326} Media Eject key.
	MEDIA_RECORD =     327, //!< {327} Media Record key.
	VOLUME_MUTE =      328, //!< {328} Speaker Mute key.
	MUSIC =            329, //!< {329} Music key.
	EXPLORER =         330, //!< {330} Explorer key.
	ENVELOPE =         331, //!< {331} Mail key.
	BOOKMARK =         332, //!< {332} Bookmark key.
	ZOOM_IN =          333, //!< {333} Zoom In key.
	ZOOM_OUT =         334, //!< {334} Zoom Out key.
}
